subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                        break 'outer;
+                    }
+                    if chunk_size == 1 {
+                        break;
+                    }
+                }
+
+                match headers_kind.get_proofs() {
+                    Ok(found) => {
+                        if found > 0 {
+                            info!("found proof {}", found)
+                        }
+                    }
+                    Err(e) => warn!("error in getting proofs {:?}", e),
+                }
+            }
+        });
+        self.closer.handles.push(headers_handle);
+
+        let mut syncer = match &self.url {
+            ElectrumUrl::Tls(url, validate) => {
+                let client = electrum_client::Client::new_ssl(url.as_str(), *validate)?;
+                SyncerKind::Tls(
+                    Syncer::new(db.clone(), client, master_blinding.clone(), self.network.clone()),
+                    url.to_string(),
+                    *validate,
+                )
+            }
+            ElectrumUrl::Plaintext(url) => {
+                let client = electrum_client::Client::new(&url)?;
+                SyncerKind::Plain(
+                    Syncer::new(db.clone(), client, master_blinding.clone(), self.network.clone()),
+                    url.to_string(),
+                )
+            }
+        };
+
+        let mut tipper = match &self.url {
+            ElectrumUrl::Tls(url, validate) => {
+                let client = electrum_client::Client::new_ssl(url.as_str(), *validate)?;
+                TipperKind::Tls(
+                    Tipper::new(db.clone(), client, self.network.clone())?,
+                    url.to_string(),
+                    *validate,
+                )
+            }
+            ElectrumUrl::Plaintext(url) => {
+                let client = electrum_client::Client::new(&url)?;
+                TipperKind::Plain(
+                    Tipper::new(db.clone(), client, self.network.clone())?,
+                    url.to_string(),
+                )
+            }
+        };
+
+        if self.wallet.is_none() {
+            let wallet = WalletCtx::new(
+                db,
+                mnemonic.clone(),
+                self.network.clone(),
+                xprv,
+                xpub,
+                master_blinding,
+            )?;
+
+            self.wallet = Some(wallet);
+        }
+
+        let notify_blocks = self.notify.clone();
+
+        let mut last_tip = tipper.tip()?;
+        info!("tip is {:?}", last_tip);
+        notify_block(notify_blocks.clone(), last_tip);
+
+        let (close_tipper, r) = channel();
+        self.closer.senders.push(close_tipper);
+        let tipper_handle = thread::spawn(move || {
+            info!("starting tipper thread");
+            loop {
+                if wait_or_close(&r, sync_interval) {
+                    info!("closing tipper thread");
+                    break;
+                }
+
+                match tipper.tip() {
```

Comment:
The tip obtained here doesn't get verified against the chain, does that matter?

Reply:
It would be nice the tip is verified against the chain, however, the headers chain is optional and one requirement was "to load fast" thus we are trusting the electrum server (the plan is to have multiple public ones or owned one)

Reply:
Perhaps it makes sense to verify it against the headers chain if the user chose to enable SPV validation?

This brings up a potential issue related to the [last comment](https://github.com/Blockstream/gdk/pull/47#discussion_r479705307) I just wrote regarding draining the chain.

What happens if a transaction gets confirmed and SPV validated at the tip (say at height N with 1 confirmation), then the server sends a fake tip message with height=N+10. Will the SPV-validated transaction show 1 or 11 confirmations?

This could be combined with draining the chain, so it'll take longer for the user to notice that something is up.

If that's indeed an issue, one way to solve that is to always use the last block in the SPV-validated headers chain as the tip height for the purpose of calculating the number of confirmations. (But then, what would the invalidated tip from `tipper` be good for? So maybe better to just always validate the tip?)

Reply:
same answer https://github.com/Blockstream/gdk/pull/47#discussion_r488471773